File: impls/zig/step1_read_print.zig
const std = @import("std");
const reader = @import("reader.zig");
const printer = @import("printer.zig");

fn READ(alloc: std.mem.Allocator, str: []const u8) !reader.Ast {
    return try reader.read_str(alloc, str);
}

fn EVAL(ast: reader.Ast) reader.Ast {
    return ast;
}

fn PRINT(ast: reader.Ast) !void {
    try printer.pr_str(ast);
}

fn rep(alloc: std.mem.Allocator, str: []const u8) !void {
    const ast = READ(alloc, str) catch |err| {
        try printer.pr_err(err);
        return;
    };
    const evalled_ast = EVAL(ast);
    try PRINT(evalled_ast);
}

pub fn main() !void {
    const stdout = std.io.getStdOut();
    const stdin = std.io.getStdIn().reader();
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    var line = std.ArrayList(u8).init(alloc);
    defer line.deinit();
    const writer = line.writer();

    try stdout.writeAll("Lispy REPL 0.0.2\n");
    try stdout.writeAll("lisp> ");
    while (stdin.streamUntilDelimiter(writer, '\n', 200)) {
        defer line.clearRetainingCapacity();
        try rep(alloc, line.items);
        try stdout.writeAll("lisp> ");
    } else |err| {
        return err;
    }
}

test "simple test" {}

File: impls/rust/src/step1_read_print.rs
use std::io;
use std::io::Write;

use printer::pr_str;
use reader::ReaderError;
use types::MalType;

mod printer;
mod reader;
mod types;

fn read(s: String) -> Result<MalType, ReaderError> {
    reader::read_str(s)
}

fn eval(input: MalType) -> MalType {
    input
}

fn print(input: MalType) -> String {
    pr_str(&input, true)
}

fn rep(s: String) -> Result<String, ReaderError> {
    Ok(print(eval(read(s)?)))
}

fn main() {
    loop {
        print!("user> ");
        io::stdout().flush().unwrap();

        let mut input = String::new();
        match io::stdin().read_line(&mut input) {
            Ok(0) => break, // empty line
            Ok(_) => {
                if input == ",q\n" {
                    break;
                }
                match rep(input) {
                    Ok(str) => println!("{}", str),
                    _ => continue,
                }
            }

            Err(_) => println!("Readline error."),
        }
    }
}
